Repeat Record Report Improvements

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
Before
![image](https://github.com/user-attachments/assets/3f530a3c-3c8b-4182-97b8-fd0dbac6e77c)

After
![image](https://github.com/user-attachments/assets/af66e1b3-f6da-49de-b4a5-bf1f71b6d2fa)

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-15977

Redo of https://github.com/dimagi/commcare-hq/pull/35782, where commits are a bit easier to review (hopefully).

Generally speaking, this page currently has multiple ways to do the same thing. For example you can resend a single payload by either clicking the button in its row, or selecting the checkbox for the row and clicking resend up above. I've modeled this changes after how gmail handles bulk actions on inboxes. As the screenshots show, I removed the dedicated options for selecting records in bulk, instead relying on using the table's select all functionality to select the _entire_ table if you want, similar to gmail.

Additionally I removed the two columns for taking actions directly on a repeat record instead relying on checking the box for that row and taking the action, which trades off a slightly more cumbersome UX for a simpler report.

The other main UX decision was to disable the bulk action buttons when the page was in a state where those actions were not allowed. Previously, we allowed the user to click the button, confirm the modal, and then would tell them the action was not allowed. Now all validation happens up front, only enabling buttons once they are actually allowed to be used.

~I plan to make a screen recording demoing the page and go through the proper comms for large UI changes. Curious to get some eyes on this before proceeding much further though.~

After discussion with Kishan, we deemed that this was not a large UI change because functionality has not changed. I've notified support of this change, but otherwise do not plan to make any general announcements.

🐟 Review by commit

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested and ran through QA. The blast radius is limited to this page only and does not impact typical repeat record processing.

This hasn't changed any functionality of this page, but just reduced the number of options users have to take virtually the same actions.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
This is mostly frontend changes that are not tested. There are a few tests for helper methods on the backend but nothing substantial.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-7596

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Remove repeat_record_report_select.js

I'm going to simplify this page quite a bit, so the main js file will
not be as daunting as it is right now. Consolidating these two js files
will make it easier to see what is happening in subsequent commits.

gherceg committed:
Replace selectItems/unSelectItems with toggleItems

gherceg committed:
Replace select all/none with simple checkbox

It is a common design pattern to have a simple checkbox in the header
row of a table select all items on the page.

gherceg committed:
Rename xform_ids to record_ids

Payloads can be more than just form ids

gherceg committed:
Rename references to action buttons

gherceg committed:
Check and validate rows based on is_queued field

Previously, validation was done by inspecting if the row had a button
that was set to requeue or cancel. I'm planning to remove that button
from the row, so this commit removes the dependency for checking and
validating rows based on that button.

gherceg committed:
Remove actions on individual payloads

This can already be accomplished by checking the box for the targeted
row, and then using the buttons above the report. Let's just stick with
one way to do things on this page.

gherceg committed:
Update all action buttons based on checked items

This also disables all action buttons by default, which is how it should
be.

gherceg committed:
Remove select all pending and cancelled buttons

Again, this can already be done by filtering the report to the specific
state you want to take action on. These buttons are also a pain to
manage as repeat record states evolve. For example, it isn't clear to a
user what a pending record is, since that could be Failed or Pending.

gherceg committed:
Replace select all with table select all checkbox

gherceg committed:
Use state from initial page data

Also generally refactor to refer to "state" where relevant instead
of "flag"

gherceg committed:
Minor refactor to action buttons

- remove 'all' from IDs since actions can apply to any number of records
- move Requeue button to before Cancel button
- change icon for Requeue (looks more like "retry" to me)

gherceg committed:
Reset table selections when checking any row

gherceg committed:
Determine current page size and update UI

Use page size to determine if select entire table button should be
displayed at all, and to make the display text more meaningful.

gherceg committed:
Reset everything if user interacts with pagination

gherceg committed:
Ran ./scripts/prettify-changed.sh

gherceg committed:
Ran ./manage.py build_bootstrap5_diffs

gherceg committed:
Fixed my bad javascript

orangejenny commented on corehq/motech/repeaters/static/repeaters/js/bootstrap3/repeat_record_report.js:
```diff
@@ -367,18 +367,10 @@ hqDefine('repeaters/js/bootstrap3/repeat_record_report', [
         });
 
-        function selectItems() {
+        function toggleItems(checked) {
```
I don't think this is an improvement, because it isn't obvious to the reader what the true/false parameter means, you have to go to the function definition to look it up.

gherceg replied:
37834d433c079d41c7d205230f0edba4f99a91a5

orangejenny commented on corehq/motech/repeaters/views/repeat_records.py:
```diff
@@ -226,13 +226,7 @@ def _make_row(self, record):
-                    _('all'),
-                    _('none')
-                ),
+                format_html('<input type="checkbox" id="select-all-checkbox"></input>'),
```
This is so much nicer.

orangejenny commented on corehq/motech/repeaters/static/repeaters/js/bootstrap3/repeat_record_report.js:
```diff
@@ -205,16 +205,15 @@ hqDefine('repeaters/js/bootstrap3/repeat_record_report', [
+        function isActionPossibleForCheckedItems(action, checkedItems) {
+            for (const item of checkedItems) {
+                const isQueued = item.getAttribute('is_queued');
+                if (isQueued == 'false' && action == 'cancel') {
```
Why is `isQueued` checking for false? It looks like the values are `0`/`1`, or possibly `"0"`/`"1"` since `getAttribute` returns a string.

orangejenny replied:
The suggestion below is the best way I can think of, although I don't love that it takes three steps, from string to int to boolean.

You can set a boolean in HTML, but you have to omit it instead of saying `my_val="false"`, which I think is less explicit/readable (I'm looking at [these docs](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML)).

gherceg replied:
Ok cool. Ended up going with your suggestion below in ea7eaf4463e40497b32e46e13256d2048bc4c4cc

orangejenny commented on corehq/motech/repeaters/static/repeaters/js/bootstrap3/repeat_record_report.js:
```diff
@@ -395,15 +394,13 @@ hqDefine('repeaters/js/bootstrap3/repeat_record_report', [
-                    } else {
-                        item.checked = true;
-                    }
+                const isQueued = item.getAttribute('is_queued');
```
You might call `parseInt(item.getAttribute('is_queued'))` here, or even `!!parseInt(item.getAttribute('is_queued'))` so you're just dealing with booleans in the rest of the code.

gherceg replied:
I realized I was making the logic more complicated than it needed to be, and also doing it differently in two different places, so this commit resolved those issues as well as used the more boolean conversion logic: ea7eaf4463e40497b32e46e13256d2048bc4c4cc

orangejenny commented on corehq/motech/repeaters/views/repeat_records.py:
```diff
@@ -71,26 +71,6 @@ def __init__(self, *args, **kwargs):
             toggles.NAMESPACE_DOMAIN,
         )
 
-    def _make_cancel_payload_button(self, record_id):
```
All of this deletion is beautiful.

orangejenny commented on corehq/motech/repeaters/templates/repeaters/bootstrap3/repeat_record_report.html:
```diff
@@ -79,6 +79,7 @@ <h4 class="modal-title">{% blocktrans %}Confirmation{% endblocktrans %}</h4>
     {% registerurl 'requeue_repeat_record' domain %}
     {% initial_page_data 'payload_id' payload_id %}
     {% initial_page_data 'repeater_id' repeater_id %}
+    {% initial_page_data 'state' state %}
```
You could grab this in js from the filter, right? I don't feel strongly about this, but I was a little surprised to see it here.

orangejenny replied:
Same comment for `total`, although it's probably a bit grosser to parse the total out of the table.

gherceg replied:
I was mainly following how it was done for payload_id and repeater_id with regards to state. The page gets re-rendered any time we apply new filters, so this felt easy enough. And generally like the idea of being handed the value rather than inferring it from the UI

orangejenny commented on corehq/motech/repeaters/static/repeaters/js/bootstrap5/repeat_record_report.js:
```diff
@@ -273,22 +275,41 @@ hqDefine('repeaters/js/bootstrap5/repeat_record_report', [
-            uncheckSelects();
+            if (selectAllCheckbox.checked) {
+                toggleItems(true);
+                selectedPageInfo.classList.remove('hide');
```
B5 doesn't have the `hide` class anymore, you can use `d-none` instead.

gherceg replied:
00533dda42639e64ec8ec5dee67ba8b56f8173cc

orangejenny reviewed (commented):
This looks good overall.

gherceg commented on corehq/motech/repeaters/static/repeaters/js/bootstrap3/repeat_record_report.js:
```diff
@@ -311,5 +313,111 @@ hqDefine('repeaters/js/bootstrap3/repeat_record_report', [
             $('#not-allowed').addClass('hide');
         }
+
+        // ----- what was once repeat_record_report_selects.js ------
```
Whoops yeah I meant to remove this 👍🏻

gherceg replied:
7ac0087e4c4b0eafcf489083218200f6611699da

gherceg committed:
Remove comment and group jquery callbacks together

gherceg committed:
B5: Use d-none instead of hide class

gherceg committed:
Replace toggleItems with explicit functions

gherceg committed:
DRY logic to check if queued records are selected

gherceg committed:
Fix javascript lint errors

One day I'll remember to add semi-colons

gherceg committed:
Always allow resend regardless of record states

gherceg committed:
Followup: action is always possible for resend

gherceg committed:
Fix linting error

gherceg committed:
Run build_bootstrap5_diffs

gherceg committed:
Format total number of records

gherceg committed:
Use record_id parameter to determine code path

State was not reliable as it didn't indicate if the entire table or
specific record ids were being selected. This change makes it so that we
look at the record_id parameter to determine if we should process an
entire queryset of repeat records based on report filters, or just a set
list of record ids passed up by the user.

gherceg committed:
Require repeater or payload for actions on table

This restores the previous behavior of preventing a bulk action on the
entire queryset if a repeater or payload is not selected. This was
previously only enforced on the backend, but is now on both.

gherceg committed:
Remove validation after bulk action button clicks

This validation is now done by enabling/disabling buttons that are
allowed based on the current state of the page, so we don't need to
validate after a button is clicked.

gherceg committed:
Reset page after performing bulk action

gherceg commented:
QA finished but there were a couple of things I noticed that needed improvements. I'll ask QA to do a brief pass on this once more, and do not plan to make any more changes to this PR unless absolutely necessary. Apologies for the churn.

gherceg committed:
Fix issues

- fix failing test for _get_state
- fix JS linting errors

orangejenny committed:
Ran hqdefine_to_esm over repeat_record_report.js

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Merge pull request #36365 from dimagi/jls/esm-repeat-records

ESM for repeat records

gherceg committed:
Use state.name instead of label

We no longer display the state field in the UI, and therefore the
non-presentable name can be used (e.g., "InvalidPayload" instead of
"Invalid Payload"). When translating state back on bulk actions
requests, we use state.name to identify the state, and therefore
require state.name be used in the first place.

gherceg committed:
Less specific button text

gherceg committed:
Merge branch 'master' into gh/repeat-record-report
